### PR TITLE
[runner] Record more metrics

### PR DESCRIPTION
**Pull request by jacek-oet**
### Runner

- Added metrics: **Max Integrality Violation** and **Duality Gap**

- **Test Cases**: Evaluated with two sample problems:
  - `runner/sample-problems/sample_lp.lp`
  - `runner/sample-problems/sample_mip.lp`

### Test Results

**Benchmark Results**

| Benchmark   | Solver | Status | Termination Condition | Objective Value | Runtime (s) | Memory Usage (MB) | Max Integrality Violation | Duality Gap |
|-------------|--------|--------|-----------------------|-----------------|-------------|-------------------|---------------------------|-------------|
| sample_lp   | gurobi | ok     | optimal               | 11.75           | 0.0047      | 162.332           | 0.25                      |             |
| sample_lp   | gurobi | ok     | optimal               | 11.75           | 0.0039      | 161.972           | 0.25                      |             |
| sample_lp   | gurobi | ok     | optimal               | 11.75           | 0.0038      | 162.096           | 0.25                      |             |
| sample_lp   | gurobi | ok     | optimal               | 11.75           | 0.0042      | 162.228           | 0.25                      |             |
| sample_lp   | highs  | ok     | optimal               | 11.75           | 0.0054      | 158.472           | 0.25                      | inf         |
| sample_lp   | highs  | ok     | optimal               | 11.75           | 0.0053      | 158.196           | 0.25                      | inf         |
| sample_lp   | highs  | ok     | optimal               | 11.75           | 0.0024      | 158.140           | 0.25                      | inf         |
| sample_lp   | highs  | ok     | optimal               | 11.75           | 0.0053      | 158.332           | 0.25                      | inf         |
| sample_lp   | glpk   | ok     | optimal               | 11.75           | 0.0092      | 157.952           | 0.25                      |             |
| sample_lp   | glpk   | ok     | optimal               | 11.75           | 0.0095      | 157.964           | 0.25                      |             |
| sample_lp   | glpk   | ok     | optimal               | 11.75           | 0.0091      | 158.220           | 0.25                      |             |
| sample_lp   | glpk   | ok     | optimal               | 11.75           | 0.0090      | 158.088           | 0.25                      |             |
| sample_lp   | scip   | ok     | optimal               | 11.75           | 0.0065      | 167.780           | 0.25                      | 0.0         |
| sample_lp   | scip   | ok     | optimal               | 11.75           | 0.0065      | 167.916           | 0.25                      | 0.0         |
| sample_lp   | scip   | ok     | optimal               | 11.75           | 0.0065      | 167.676           | 0.25                      | 0.0         |
| sample_lp   | scip   | ok     | optimal               | 11.75           | 0.0066      | 167.792           | 0.25                      | 0.0         |
| sample_mip  | gurobi | ok     | optimal               | 30.00           | 0.0038      | 162.644           | 0.0                       | 0.0         |
| sample_mip  | gurobi | ok     | optimal               | 30.00           | 0.0042      | 162.392           | 0.0                       | 0.0         |
| sample_mip  | gurobi | ok     | optimal               | 30.00           | 0.0039      | 162.516           | 0.0                       | 0.0         |
| sample_mip  | gurobi | ok     | optimal               | 30.00           | 0.0040      | 162.344           | 0.0                       | 0.0         |
| sample_mip  | highs  | ok     | optimal               | 30.00           | 0.0045      | 158.284           | 0.0                       | 0.0         |
| sample_mip  | highs  | ok     | optimal               | 30.00           | 0.0043      | 158.184           | 0.0                       | 0.0         |
| sample_mip  | highs  | ok     | optimal               | 30.00           | 0.0053      | 158.412           | 0.0                       | 0.0         |
| sample_mip  | highs  | ok     | optimal               | 30.00           | 0.0023      | 158.248           | 0.0                       | 0.0         |
| sample_mip  | glpk   | ok     | optimal               | 30.00           | 0.0091      | 157.964           | 0.0                       |             |
| sample_mip  | glpk   | ok     | optimal               | 30.00           | 0.0093      | 157.976           | 0.0                       |             |
| sample_mip  | glpk   | ok     | optimal               | 30.00           | 0.0090      | 157.868           | 0.0                       |             |
| sample_mip  | glpk   | ok     | optimal               | 30.00           | 0.0090      | 158.236           | 0.0                       |             |
| sample_mip  | scip   | ok     | optimal               | 30.00           | 0.0173      | 167.812           | 0.0                       | 0.0         |
| sample_mip  | scip   | ok     | optimal               | 30.00           | 0.0086      | 167.564           | 0.0                       | 0.0         |
| sample_mip  | scip   | ok     | optimal               | 30.00           | 0.0109      | 167.816           | 0.0                       | 0.0         |
| sample_mip  | scip   | ok     | optimal               | 30.00           | 0.0142      | 167.420           | 0.0                       | 0.0         |

---

**Benchmark Mean and StdDev Results**

| Benchmark   | Solver | Status | Termination Condition | Objective Value | Runtime Mean (s) | Runtime StdDev (s) | Memory Mean (MB) | Memory StdDev (MB) |
|-------------|--------|--------|-----------------------|-----------------|------------------|--------------------|------------------|--------------------|
| sample_lp   | gurobi | ok     | optimal               | 11.75           | 0.0041           | 0.0004             | 162.157          | 0.157              |
| sample_lp   | highs  | ok     | optimal               | 11.75           | 0.0046           | 0.0014             | 158.285          | 0.148              |
| sample_lp   | glpk   | ok     | optimal               | 11.75           | 0.0092           | 0.0002             | 158.056          | 0.125              |
| sample_lp   | scip   | ok     | optimal               | 11.75           | 0.0065           | 0.00005            | 167.791          | 0.098              |
| sample_mip  | gurobi | ok     | optimal               | 30.00           | 0.0040           | 0.00015            | 162.474          | 0.135              |
| sample_mip  | highs  | ok     | optimal               | 30.00           | 0.0041           | 0.0013             | 158.282          | 0.096              |
| sample_mip  | glpk   | ok     | optimal               | 30.00           | 0.0091           | 0.00015            | 158.011          | 0.158              |
| sample_mip  | scip   | ok     | optimal               | 30.00           | 0.0127           | 0.0038             | 167.653          | 0.195              |

---

### Command

```bash
python runner/run_benchmarks.py
```

### Website
Add a new table for MIP Information (displayed only for MIP problems) on the benchmark detail page.

image

Note: I assume pypsa-gas+wind+sol+ely-1h-ucgas and pypsa-gas+wind+sol+ely-1h are MIP problems, so I added mock data for them (Max Integrality Violation, Duality Gap).

### CODE REVIEW

**Comment by siddharth-krishna** [comment]
Thanks Jacek! This is a good start. Some suggestions for next steps:

- [ ] Check if the code also works for other solvers (glpk, scip)?
- [ ] Add `max_integrality_violation` and `duality_gap` to `results` returned by `run_solver.py`
- [ ] What does linopy return for the above 2 metrics when the input file is not a MIP? If it is something sensible, then run this code on all our benchmarks. Otherwise, can we figure out a way to ask linopy / `run_benchmarks.py` whether each benchmark is an LP or MIP and only run the above code for MIPs?

**Comment by jacek-oet** [comment]
The previous code only works for the HiGHS solver. Other solvers use a different method to retrieve those values.
I updated the code to get max_integrality_violation for all solvers. I am now checking how to retrieve the duality_gap for all solvers and I will run all benchmarks.

File: runner/run_solver.py [hunk 1]
```diff
@@ -42,28 +44,18 @@ def main(solver_name, input_file):
...
-        if max_integrality_violation == 0.0:
-            print("Integer feasible solution found.")
-        else:
-            print("Integer feasible solution not found.")
-
-        if mip_gap == 0.0:
-            print("Optimal solution found with a duality gap of 0%.")
-        else:
-            print(f"Non-zero duality gap: {mip_gap * 100:.2f}%")
+    primal_values = solver_result.solution.primal
```
Comment (siddharth-krishna) [inline runner/run_solver.py:47]: Thanks! Please put a comment here that we are not using `solver_result.solver_model.getInfo()` because it works for Highs but not solver X

**Comment by jacek-oet** [comment]
I added `max_integrality_violation` and `duality_gap` to the results returned by `run_solver.py` and am re-running the code on all our benchmarks. However, for GLPK, I can't get the `duality_gap` because, in Linopy, it only returns `status` and `solution` (`Result(status, solution)`). Other solvers return a model, which allows us to retrieve the `duality_gap` from it.

> For LP, if we use IPM, do we crossover to get a basic solution? What solution quality do we require? Vertex or optimal?

For linear programming (LP), if we use an interior point method (IPM), the process does not naturally yield a basic (vertex) solution because IPMs find points within the feasible region rather than at its vertices.

To obtain a basic (vertex) solution from an IPM, a crossover step is typically used. This step takes the approximate solution from the IPM and moves it to a basic feasible solution, ideally at a vertex of the feasible region. This is particularly useful in cases where a vertex (or extreme point) is required.

In summary, the solution quality primarily requires optimality and feasibility, but if a vertex solution is specifically needed, a crossover step is necessary.

**Review by siddharth-krishna — COMMENTED** [review]
Thanks, Jacek. We also need to add these new metrics to the website. For now, perhaps the simplest way to do this is to add a little table to the Benchmark details page, with columns "Solver, Max Integrality Violation, Duality Gap". You can display the results from the last iteration for each solver, if the results CSV has multiple iterations. Also, make sure you only display this info for Benchmarks that are MIPs, not for LPs. Thanks!

**Comment by jacek-oet** [comment]
I added a new table for MIP Information (displayed only for MIP problems) on the benchmark detail page.
image

Note: I assume pypsa-gas+wind+sol+ely-1h-ucgas and pypsa-gas+wind+sol+ely-1h are MIP problems, so I added mock data for them (Max Integrality Violation, Duality Gap).

**Comment by siddharth-krishna** [comment]
Thanks Jacek. The table looks good. But I think we want to programmatically determine which benchmarks are MIPs. Can we use the `Technique:` line from `benchmarks/pypsa/metadata.yaml`?

Also, I don't see the code updating the website in the PR's diff. Did you not push it yet? Thanks

**Comment by jacek-oet** [comment]
> Thanks Jacek. The table looks good. But I think we want to programmatically determine which benchmarks are MIPs. Can we use the `Technique:` line from `benchmarks/pypsa/metadata.yaml`?
> 
> Also, I don't see the code updating the website in the PR's diff. Did you not push it yet? Thanks

I will use the Technique field from metadata.yaml and add mock data accordingly

File: runner/run_solver.py [hunk 2]
```diff
@@ -30,23 +31,74 @@ def get_solver(solver_name):
...
+
+
 def main(solver_name, input_file):
     problem_file = Path(input_file)
     solver = get_solver(solver_name)
-    solution_fn = Path(f"runner/solution/{problem_file.stem}-{solver_name}.sol")
+    solution_dir = Path(__file__).parent / "solution"
+    solution_dir.mkdir(parents=True, exist_ok=True)
+    solution_fn = (
+        Path(__file__).parent / "solution/{problem_file.stem}-{solver_name}.sol"
```
Comment (siddharth-krishna) [inline runner/run_solver.py:75]: why not use `solution_dir` in the path here?

File: runner/run_solver.py [hunk 3]
```diff
@@ -30,23 +31,74 @@ def get_solver(solver_name):
...
     )
     runtime = time() - start_time
+    solver_model = solver_result.solver_model
+    duality_gap = None
+    max_integrality_violation = None
+
+    if is_mip_problem(solver_model, solver_name):
+        duality_gap = get_duality_gap(solver_model, solver_name)
+        max_integrality_violation = calculate_integrality_violation(
+            # We are not using solver_result.solver_model.getInfo() because it works for HiGHS but not for other solvers.
```
Comment (siddharth-krishna) [inline runner/run_solver.py:91]: Perhaps now this comment makes more sense to be put inside the `calculate_integrality_violation` function

File: website/benchmarks.py [hunk 4]
```diff
@@ -133,5 +137,34 @@ def load_metadata(yaml_file):
...
+    if metadata[selected_benchmark].get("Technique") == "MILP":
+        mip_data = df_result[(df_result["Benchmark"] == selected_benchmark)]
+
+        if not mip_data.empty:
+            st.subheader(f"MIP Information for {selected_benchmark}")
+
+            # Display MIP-specific information in a table
+            mip_table = mip_data[
+                ["Solver", "Max Integrality Violation", "Duality Gap"]
+            ].drop_duplicates()
```
Comment (siddharth-krishna) [inline website/benchmarks.py:154]: Not sure if this would work? What if different iterations of a solver had different duality gaps for some reason? Perhaps better to drop duplicates by only considering the solver column, or to consider the last row for each solver?

File: runner/run_solver.py [hunk 5]
```diff
@@ -30,23 +31,74 @@ def get_solver(solver_name):
...
+def is_mip_problem(solver_model, solver_name):
+    if solver_name == "scip":
+        if solver_model.getNIntVars() > 0 or solver_model.getNBinVars() > 0:
+            return True
+        return False
+    elif solver_name == "gurobi":
+        return solver_model.IsMIP
+    elif solver_name == "highs":
+        info = solver_model.getInfo()
+        return info.mip_node_count >= 0
```
Comment (siddharth-krishna) [inline runner/run_solver.py:43]: @jacek-oet what happens if the solver is GLPK?

Comment (siddharth-krishna) [reply runner/run_solver.py:43]: Perhaps a good idea in general to have a default case where it raises a NotImplementedError

**Review by siddharth-krishna — CHANGES_REQUESTED** [review]
Would love a few small changes, thanks!

File: runner/run_solver.py (hunk 5 above)
Comment (jacek-oet) [reply runner/run_solver.py:43]: If we raise a NotImplementedError, we won’t be able to use GLPK because it will throw that error every time we try to use GLPK.

Comment (siddharth-krishna) [reply runner/run_solver.py:43]: Yes, but my point is that there is no `else:` block here, which is bad coding practice. Because then perhaps the code returns `None` in the case when `solver_name == "glpk"`? It is better to add an `elif solver_name == "glpk":` block to specify what happens for GLPK, and have `else: raise NotImplementedError` to remind ourselves to update this function when we add a new solver to the code.

File: runner/run_solver.py [hunk 6]
```diff
@@ -41,10 +41,15 @@ def is_mip_problem(solver_model, solver_name):
...
         return info.mip_node_count >= 0
+    else:
+        raise NotImplementedError(f"The solver '{solver_name}' is not supported.")
 
 
 def calculate_integrality_violation(primal_values) -> float:
-    """Calculate the maximum integrality violation from primal values."""
+    """Calculate the maximum integrality violation from primal values.
+    Note:
+        We are not using solver_result.solver_model.getInfo() because it works for HiGHS but not for other solve
```
Comment (siddharth-krishna) [inline runner/run_solver.py:51]:
```suggestion
        We are not using solver_result.solver_model.getInfo() because it works for HiGHS but not for other solvers
```

**Review by siddharth-krishna — APPROVED** [review]
